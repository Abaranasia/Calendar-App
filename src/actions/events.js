import Swal from "sweetalert2";
import { fetchConToken } from "../helpers/fetch"
import { types } from "../types/types"


export const eventStartAddNew = (event) => {
    return async (dispatch, getState) => {

        const { uid, name } = getState().auth;

        try {
            const resp = await fetchConToken('events', event, 'POST');
            const body = await resp.json();
            console.log(body)
            if (body.ok) {
                event.id = body.evento.id;
                event.user = { // el usuario lo podemos sacar del state de redux con getState
                    _id: uid,
                    name: name
                }

                dispatch(eventAddNew(event))
            }
        } catch (error) {
            console.log(error)
            Swal.fire('Error', error, 'error')
        }
    }

}

const eventAddNew = (event) => ({ // AÃ±adir nuevo evento, pero solo al guardar en la BBDD. Sin export por ser interna
    type: types.eventAddNew,
    payload: event
})


export const eventSetActive = (event) => ({ // Establecer como activo un evento (y mostrarlo en el modal)
    type: types.eventSetActive,
    payload: event
})


export const eventClearActiveNote = () => ({ // Elimina el evento activo
    type: types.eventClearActiveEvent
})


export const eventUpdated = (event) => ({ // permite actualizar un evento existente
    type: types.eventUpdated,
    payload: event
})


export const eventDeleted = () => ({ // permite eliminar un evento existente
    type: types.eventDeleted
})