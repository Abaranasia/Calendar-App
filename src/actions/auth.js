import Swal from "sweetalert2";
import { fetchConToken, fetchSinToken } from "../helpers/fetch";
import { types } from "../types/types";
import { eventLogout } from "./events";


export const startLogin = (email, password) => {
    return async (dispatch) => { // Pongo este return porque es una tarea asÃ­ncrona. Dispatch proporcionado por thunk

        // console.log("AUTH: ", email, password);
        const resp = await fetchSinToken('auth', { email, password }, 'POST'); // Nuestro helper sin token
        const body = await resp.json();
        // console.log(body)

        if (body.ok) {
            localStorage.setItem('token', body.token)
            localStorage.setItem('token-init-date', new Date().getTime()) // Momento de inicio del token

            dispatch(login({ // Loguea en redux
                uid: body.uid,
                name: body.name
            }))
        } else {
            Swal.fire('Error', body.msg, 'error')
        }

    }
}


export const startRegister = (name, email, password) => {
    return async (dispatch) => {
        const resp = await fetchSinToken('auth/new', { name, email, password }, 'POST'); // Nuestro helper sin token
        const body = await resp.json();
        // console.log(body)

        if (body.ok) {
            localStorage.setItem('token', body.token)
            localStorage.setItem('token-init-date', new Date().getTime()) // Momento de inicio del token

            dispatch(login({ // Loguea en redux
                uid: body.uid,
                name: body.name
            }))
        } else {
            Swal.fire('Error', body.msg, 'error')
        }
    }
}


export const startChecking = () => { // Comprobamos que se ha renovado el token
    return async (dispatch) => {
        const resp = await fetchConToken('auth/renew', {}, 'GET'); // Nuestro helper sin token
        const body = await resp.json();
        console.log(body)

        if (body.ok) {
            localStorage.setItem('token', body.token)
            localStorage.setItem('token-init-date', new Date().getTime()) // Momento de inicio del token

            dispatch(login({ // Loguea en redux
                uid: body.uid,
                name: body.name
            }))
        } else {
            dispatch(checkingFinish())
            Swal.fire('Error', body.msg, 'error');
        }
    }
}


export const startLogout = () => {
    return (dispatch) => {
        localStorage.clear();
        dispatch(eventLogout()); // limpia evento activo
        dispatch(eventLogout()); // limpia evento activo
        dispatch(logout());  //limpiar auth y localStorage
    }
}


const checkingFinish = () => ({ // Action para terminar el checking del token
    type: types.authCheckingFinish
})

const login = (user) => ({ // Action para login
    type: types.authLogin,
    payload: user
})

const logout = () => ({
    type: types.authLogout
})
