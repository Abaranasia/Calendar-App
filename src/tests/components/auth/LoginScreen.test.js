import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';

import Swal from 'sweetalert2';

import { LoginScreen } from '../../../components/auth/LoginScreen';
import { startLogin, startRegister } from '../../../actions/auth';



jest.mock('../../../actions/auth', () => ({
    startLogin: jest.fn(),
    startRegister: jest.fn()
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};
const store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <LoginScreen />
    </Provider>
)

describe('<LoginScreen /> tests', () => {


    beforeEach(() => {
        jest.clearAllMocks()
    })


    test('should display properly', () => {

        expect(wrapper).toMatchSnapshot()
    });



    test('should call login dispatch', () => {

        wrapper.find('input[name="loginEmail"]').simulate('change', {
            target: {
                name: 'loginEmail',
                value: 'test@testing.com'
            }
        })

        wrapper.find('input[name="loginPassword"]').simulate('change', {
            target: {
                name: 'loginPassword',
                value: '1234567'
            }
        })

        wrapper.find('form').at(0).prop('onSubmit')({ //  Posici칩n 0 para el form de login
            preventDefault() { }
        })

        expect(startLogin).toHaveBeenCalledWith('test@testing.com', '1234567')
    });



    test('testing different passwords', () => {

        // Simular 2 passwords diferentes
        wrapper.find('input[name="registerPassword1"]').simulate('change', {
            target: {
                name: 'registerPassword1',
                value: '1234567'
            }
        })

        wrapper.find('input[name="registerPassword2"]').simulate('change', {
            target: {
                name: 'registerPassword2',
                value: '12345'
            }
        })

        wrapper.find('form').at(1).prop('onSubmit')({ // Posici칩n 1 para el form de registro
            preventDefault() { }
        })

        expect(startRegister).not.toHaveBeenCalled();
        expect(Swal.fire).toHaveBeenCalledWith('Error', 'Las contrase침as deben ser iguales', 'error');
    });


    test('startRegister should be called with same register passwords ', () => {
        wrapper.find('input[name="registerPassword1"]').simulate('change', {
            target: {
                name: 'registerPassword1',
                value: '1234567'
            }
        })

        wrapper.find('input[name="registerPassword2"]').simulate('change', {
            target: {
                name: 'registerPassword2',
                value: '1234567'
            }
        })

        wrapper.find('form').at(1).prop('onSubmit')({ // Posici칩n 1 para el form de registro
            preventDefault() { }
        })

        expect(startRegister).toHaveBeenCalled();
        expect(Swal.fire).not.toHaveBeenCalled();
    });
})
