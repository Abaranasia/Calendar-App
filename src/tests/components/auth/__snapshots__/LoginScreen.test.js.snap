// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginScreen /> tests should display properly 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Ingreso
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="text"
            value="hhansolo@gmail.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="1234567"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerName"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value="Pepo"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value="pepo@gmail.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerPassword1"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="1234567"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerPassword2"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value="1234567"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
